v2/inc/c_logging/log_context.h

```diff
@@ -4,6 +4,241 @@
+#include "c_logging/log_context_property_type_struct.h"
+#include "c_logging/log_internal_error.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define LOG_MAX_STACK_DATA_SIZE                 1024
+#define LOG_MAX_STACK_PROPERTY_VALUE_PAIR_COUNT 64
+
+typedef struct LOG_CONTEXT_TAG
+{
+    uint8_t* values_data;
+    uint32_t values_data_length;
+    LOG_CONTEXT_PROPERTY_VALUE_PAIR* property_value_pairs_ptr;
+    uint32_t property_value_pair_count;
+    LOG_CONTEXT_PROPERTY_VALUE_PAIR property_value_pairs[];
+} LOG_CONTEXT;
+
 typedef struct LOG_CONTEXT_TAG* LOG_CONTEXT_HANDLE;
 
+uint32_t log_context_get_property_value_pair_count(LOG_CONTEXT_HANDLE log_context);
+const LOG_CONTEXT_PROPERTY_VALUE_PAIR* log_context_get_property_value_pairs(LOG_CONTEXT_HANDLE log_context);
+
+// These is an internal API and it is not meant to be called by the users of this module
+int internal_log_context_init_from_parent(LOG_CONTEXT_HANDLE dest_log_context, LOG_CONTEXT_HANDLE parent_log_context);
+uint32_t internal_log_context_get_values_data_length_or_zero(LOG_CONTEXT_HANDLE log_context);
+
+static uint32_t internal_log_context_get_property_value_pair_count_or_zero(LOG_CONTEXT_HANDLE log_context)
+{
+    uint32_t result = log_context_get_property_value_pair_count(log_context);
+    if (result == UINT32_MAX)
+    {
+        result = 0;
+    }
+    return result;
+}
+
+// macro set used to define a parameter in a function signature in order
+// to make sure that no properties with the same name are added in one context
+
+// DEFINE_PROPERTY_AS_PARAMETER
+
+#define EXPAND_DEFINE_PROPERTY_AS_PARAMETER_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+    , int property_name
+
+#define EXPAND_DEFINE_PROPERTY_AS_PARAMETER_LOG_CONTEXT_NAME(log_context_name) \
+
+#define EXPAND_DEFINE_PROPERTY_AS_PARAMETER_LOG_CONTEXT_PROPERTY(property_type, property_name, field_value) \
+    , int property_name
+
+#define DEFINE_PROPERTY_AS_PARAMETER(field_desc) \
+    MU_C2(EXPAND_DEFINE_PROPERTY_AS_PARAMETER_, field_desc)
+
+// macro set used to define a parameter in a function signature  every time LOG_CONTEXT_NAME is used.
+// if used twice a field with the same name is defined => compiler error
+
+// DEFINE_LOG_CONTEXT_NAME_AS_FIELD
+
+#define EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+
+#define EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_LOG_CONTEXT_NAME(log_context_name) \
+    , int log_context_is_used_multiple_times
+
+#define EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_LOG_CONTEXT_PROPERTY(property_type, property_name, field_value) \
+
+#define DEFINE_CONTEXT_NAME_AS_PARAMETER(field_desc) \
+    MU_C2(EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_, field_desc)
+
+#define LOG_CONTEXT_CHECK_VARIABLE_ARGS(...) \
+    /* Codes_SRS_LOG_CONTEXT_01_019: [ If 2 properties have the same name for a context a compiler error shall be emitted. ]*/ \
+    (void)(void (*)(int MU_FOR_EACH_1(DEFINE_PROPERTY_AS_PARAMETER, __VA_ARGS__)))0x4242; \
```

Comment:
>int [](http://example.com/codeflow?start=20&length=3)

should this parameter have a name, just so I don't think this is followed by a name...

Reply:
I left it without a name so that there are 0 collisions. In this case I can afford it.